[bug] [android] [windows] [cli] [build] Android Build Error: A problem occurred starting process 'command 'C:\nvm4w\nodejs\node.exe.cmd'' and Error: Cannot find module '...tauri' when using nvm4w

### Describe the bug

Hi everyone. I am a student working on my year 3 project useing tauri and I encoutered this issue:

Every time I attempt to start a new Tauri project and run tauri android dev on Windows, I consistently encounter a build failure. The primary errors indicate an issue with finding and executing Node.js via C:\nvm4w\nodejs\node.exe.cmd, followed by a Cannot find module '...tauri' error, and ultimately a Gradle build failure on the rustBuildX86_64Debug task.

ERROR LOGS: 
-----------------------------------------------------------------------------------------------------------------------
            Finished `dev` profile [unoptimized + debuginfo] target(s) in 10.34s
                Info symlinking lib "E:\\uni\\year 3\\year 3 project\\CalendarAssistantApp\\CATdesktop\\src-tauri\\target\\x86_64-linux-android\\debug\\libapp_lib.so" in jniLibs dir "E:\\uni\\year 3\\year 3 project\\CalendarAssistantApp\\CATdesktop\\src-tauri\\gen/android\\app/src/main/jniLibs/x86_64"
                Info "E:\\uni\\year 3\\year 3 project\\CalendarAssistantApp\\CATdesktop\\src-tauri\\target\\x86_64-linux-android\\debug\\libapp_lib.so" requires shared lib "libandroid.so"
                Info "E:\\uni\\year 3\\year 3 project\\CalendarAssistantApp\\CATdesktop\\src-tauri\\target\\x86_64-linux-android\\debug\\libapp_lib.so" requires shared lib "libdl.so"
                Info "E:\\uni\\year 3\\year 3 project\\CalendarAssistantApp\\CATdesktop\\src-tauri\\target\\x86_64-linux-android\\debug\\libapp_lib.so" requires shared lib "liblog.so"
                Info "E:\\uni\\year 3\\year 3 project\\CalendarAssistantApp\\CATdesktop\\src-tauri\\target\\x86_64-linux-android\\debug\\libapp_lib.so" requires shared lib "libm.so"
                Info "E:\\uni\\year 3\\year 3 project\\CalendarAssistantApp\\CATdesktop\\src-tauri\\target\\x86_64-linux-android\\debug\\libapp_lib.so" requires shared lib "libc.so"
                Info symlink at "E:\\uni\\year 3\\year 3 project\\CalendarAssistantApp\\CATdesktop\\src-tauri\\gen/android\\app/src/main/jniLibs/arm64-v8a\\libapp_lib.so" points to "E:\\uni\\year 3\\year 3 project\\CalendarAssistantApp\\CATdesktop\\src-tauri\\target\\aarch64-linux-android\\release\\libapp_lib.so"
                Info symlink at "E:\\uni\\year 3\\year 3 project\\CalendarAssistantApp\\CATdesktop\\src-tauri\\gen/android\\app/src/main/jniLibs/x86_64\\libapp_lib.so" points to "E:\\uni\\year 3\\year 3 project\\CalendarAssistantApp\\CATdesktop\\src-tauri\\target\\x86_64-linux-android\\debug\\libapp_lib.so"
        node:internal/modules/cjs/loader:1146
          throw err;
          ^
        
        Error: Cannot find module 'E:\uni\year 3\year 3 project\CalendarAssistantApp\CATdesktop\src-tauri\tauri'
            at Module._resolveFilename (node:internal/modules/cjs/loader:1143:15)
            at Module._load (node:internal/modules/cjs/loader:984:27)
            at Function.executeUserEntryPoint [as runMain] (node:internal/modules/run_main:135:12)
            at node:internal/main/run_main_module:28:49 {
          code: 'MODULE_NOT_FOUND',
          requireStack: []
        }
        
        Node.js v20.12.2
        
        FAILURE: Build failed with an exception.
        
        Deprecated Gradle features were used in this build, making it incompatible with Gradle 9.0.
        
        You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.
        
        For more on this, please refer to https://docs.gradle.org/8.14.3/userguide/command_line_interface.html#sec:command_line_warnings in the Gradle documentation.
        
        BUILD FAILED in 2s
        Failed to assemble APK: command ["E:\\uni\\year 3\\year 3 project\\CalendarAssistantApp\\CATdesktop\\src-tauri\\gen/android\\gradlew.bat", "--project-dir", "E:\\uni\\year 3\\year 3 project\\CalendarAssistantApp\\CATdesktop\\src-tauri\\gen/android"] exited with code 1: command ["E:\\uni\\year 3\\year 3 project\\CalendarAssistantApp\\CATdesktop\\src-tauri\\gen/android\\gradlew.bat", "--project-dir", "E:\\uni\\year 3\\year 3 project\\CalendarAssistantApp\\CATdesktop\\src-tauri\\gen/android"] exited with code 1
               Error Failed to assemble APK: command ["E:\\uni\\year 3\\year 3 project\\CalendarAssistantApp\\CATdesktop\\src-tauri\\gen/android\\gradlew.bat", "--project-dir", "E:\\uni\\year 3\\year 3 project\\CalendarAssistantApp\\CATdesktop\\src-tauri\\gen/android"] exited with code 1: command ["E:\\uni\\year 3\\year 3 project\\CalendarAssistantApp\\CATdesktop\\src-tauri\\gen/android\\gradlew.bat", "--project-dir", "E:\\uni\\year 3\\year 3 project\\CalendarAssistantApp\\CATdesktop\\src-tauri\\gen/android"] exited with code 1
-----------------------------------------------------------------------------------------------------------------------

after some digging and debugging I found that this seems to be because in autogenerated Gradle BuildTask.kt it throws a specific error in windows specificaly this is the file: 

-----------------------------------------------------------------------------------------------------------------------
    import java.io.File
    import org.apache.tools.ant.taskdefs.condition.Os
    import org.gradle.api.DefaultTask
    import org.gradle.api.GradleException
    import org.gradle.api.logging.LogLevel
    import org.gradle.api.tasks.Input
    import org.gradle.api.tasks.TaskAction
    
    open class BuildTask : DefaultTask() {
        @Input
        var rootDirRel: String? = null
        @Input
        var target: String? = null
        @Input
        var release: Boolean? = null
    
        @TaskAction
        fun assemble() {
            val executable = """C:\nvm4w\nodejs\node.exe"""; // Or similar path
            try {
                runTauriCli(executable)
            } catch (e: Exception) {
                if (Os.isFamily(Os.FAMILY_WINDOWS)) {
                    runTauriCli("$executable.cmd") // This part failed as node.exe.cmd didn't exist
                } else {
                    throw e;
                }
            }
        }
    
        fun runTauriCli(executable: String) {
            // ... (rest of the function)
            val args = listOf("tauri", "android", "android-studio-script"); // These args were passed to the Node.js executable
            // ... (rest of the function)
        }
    }
-----------------------------------------------------------------------------------------------------------------------

my solution was to change the file to this: 

-----------------------------------------------------------------------------------------------------------------------
    import java.io.File
    import org.apache.tools.ant.taskdefs.condition.Os
    import org.gradle.api.DefaultTask
    import org.gradle.api.GradleException
    import org.gradle.api.logging.LogLevel
    import org.gradle.api.tasks.Input
    import org.gradle.api.tasks.TaskAction
    
    open class BuildTask : DefaultTask() {
        @Input
        var rootDirRel: String? = null
        @Input
        var target: String? = null
        @Input
        var release: Boolean? = null
    
        @TaskAction
        fun assemble() {
            // The executable should be 'cargo'
            val cargoExecutable = if (Os.isFamily(Os.FAMILY_WINDOWS)) "cargo.exe" else "cargo" // Ensures 'cargo' is used
            runTauriCli(cargoExecutable)
        }
    
        fun runTauriCli(executable: String) {
            val rootDirRel = rootDirRel ?: throw GradleException("rootDirRel cannot be null")
            val target = target ?: throw GradleException("target cannot be null")
            val release = release ?: throw GradleException("release cannot be null")
    
            // Arguments for 'cargo tauri'
            val args = mutableListOf(
                "tauri",
                "android",
                "android-studio-script",
                "--target", target
            )
    
            if (release) {
                args.add("--release")
            }
    
            if (project.logger.isEnabled(LogLevel.DEBUG)) {
                args.add("-vv")
            } else if (project.logger.isEnabled(LogLevel.INFO)) {
                args.add("-v")
            }
    
            project.exec {
                workingDir(File(project.projectDir, rootDirRel))
                executable(executable) // This will now be "cargo"
                args(args)
            }.assertNormalExitValue()
        }
    }

-----------------------------------------------------------------------------------------------------------------------
and this fixed the problem. As sstated I am a student and I dont know if this is a known issue or not but because it was showing every time I wanted to build an android app even clean tauri+vue apps I figured I share this to mybe help somebody

### Expected behavior

The tauri android dev command should successfully compile and run the application on the detected Android device/emulator without manual intervention to the generated Gradle build files.

Actual Behavior
The build fails with errors indicating that Node.js cannot be found or invoked correctly by the Gradle BuildTask, specifically looking for node.exe.cmd and then misinterpreting the tauri command as a Node.js script.

### Full `tauri info` output

```text
> catdesktop@0.0.0 tauri
> tauri info


[✔] Environment
    - OS: Windows 10.0.22631 x86_64 (X64)
    ✔ WebView2: 138.0.3351.95
    ✔ MSVC: Visual Studio Community 2022
    ✔ rustc: 1.87.0 (17067e9ac 2025-05-09)
    ✔ cargo: 1.87.0 (99624be96 2025-05-06)
    ✔ rustup: 1.28.2 (e4f3ad6f8 2025-04-28)
    ✔ Rust toolchain: stable-x86_64-pc-windows-msvc (default)
    - node: 20.12.2
    - npm: 10.5.0

[-] Packages
    - tauri 🦀: 2.7.0
    - tauri-build 🦀: 2.3.1
    - wry 🦀: 0.52.1
    - tao 🦀: 0.34.0
    - tauri-cli 🦀: 2.7.1
    - @tauri-apps/api : 2.7.0
    - @tauri-apps/cli : 2.7.1

[-] Plugins
    - tauri-plugin-log 🦀: 2.6.0
    - @tauri-apps/plugin-log : not installed!

[-] App
    - build-type: bundle
    - CSP: unset
    - frontendDist: ../dist
    - devUrl: http://localhost:5173/
    - framework: Vue.js
    - bundler: Vite

--- a/.changes/android-init-windows-unix.md
+++ b/.changes/android-init-windows-unix.md
@@ -1,0 +1,6 @@
+---
+'tauri-cli': 'patch:bug'
+'@tauri-apps/cli': patch:bug
+---
+
+Strip Windows-only extensions from the binary path so an Android project initialized on Windows can be used on UNIX systems.

--- a/.changes/fix-android-build-nvm4w-windows.md
+++ b/.changes/fix-android-build-nvm4w-windows.md
@@ -1,5 +1,6 @@
 ---
 'tauri-cli': 'patch:bug'
+'@tauri-apps/cli': patch:bug
 ---
 
-Fix Android build error on Windows when using nvm4w (issue documented at #13892 [bug] [android] [windows] [cli] [build] Android Build Error: A problem occurred starting process 'command 'C:\nvm4w\nodejs\node.exe.cmd'' and Error: Cannot find module '...tauri' when using nvm4w). The BuildTask.kt template now includes robust fallback logic for Windows executable detection, preventing the "Cannot find module" and "node.exe.cmd" errors that occurred with nvm4w Node.js installations. The fix tries multiple Windows-specific fallbacks including .cmd, .bat extensions and cargo as a last resort.
+Enhance Android build script usage on Windows by attempting to run cmd, bat and exe formats.

--- a/.changes/fix-android-build-nvm4w-windows.md
+++ b/.changes/fix-android-build-nvm4w-windows.md
@@ -1,0 +1,5 @@
+---
+'tauri-cli': 'patch:bug'
+---
+
+Fix Android build error on Windows when using nvm4w (issue documented at #13892 [bug] [android] [windows] [cli] [build] Android Build Error: A problem occurred starting process 'command 'C:\nvm4w\nodejs\node.exe.cmd'' and Error: Cannot find module '...tauri' when using nvm4w). The BuildTask.kt template now includes robust fallback logic for Windows executable detection, preventing the "Cannot find module" and "node.exe.cmd" errors that occurred with nvm4w Node.js installations. The fix tries multiple Windows-specific fallbacks including .cmd, .bat extensions and cargo as a last resort.

--- a/crates/tauri-cli/src/mobile/init.rs
+++ b/crates/tauri-cli/src/mobile/init.rs
@@ -117,7 +117,16 @@ pub fn exec(
   build_args.push(target.command_name());
   build_args.push(target.ide_build_script_name());
 
-  map.insert("tauri-binary", binary.to_string_lossy());
+  let mut binary = binary.to_string_lossy().to_string();
+  if binary.ends_with(".exe") || binary.ends_with(".cmd") || binary.ends_with(".bat") {
+    // remove Windows-only extension
+    binary.pop();
+    binary.pop();
+    binary.pop();
+    binary.pop();
+  }
+
+  map.insert("tauri-binary", binary);
   map.insert("tauri-binary-args", &build_args);
   map.insert("tauri-binary-args-str", build_args.join(" "));
 

--- a/crates/tauri-cli/templates/mobile/android/buildSrc/src/main/kotlin/BuildTask.kt
+++ b/crates/tauri-cli/templates/mobile/android/buildSrc/src/main/kotlin/BuildTask.kt
@@ -21,7 +21,24 @@ open class BuildTask : DefaultTask() {
             runTauriCli(executable)
         } catch (e: Exception) {
             if (Os.isFamily(Os.FAMILY_WINDOWS)) {
-                runTauriCli("$executable.cmd")
+                // Try different Windows-specific fallbacks
+                val fallbacks = listOf(
+                    "$executable.cmd",
+                    "$executable.bat",
+                    "cargo.exe",
+                    "cargo"
+                )
+                
+                var lastException: Exception = e
+                for (fallback in fallbacks) {
+                    try {
+                        runTauriCli(fallback)
+                        return
+                    } catch (fallbackException: Exception) {
+                        lastException = fallbackException
+                    }
+                }
+                throw lastException
             } else {
                 throw e;
             }
@@ -32,7 +49,15 @@ open class BuildTask : DefaultTask() {
         val rootDirRel = rootDirRel ?: throw GradleException("rootDirRel cannot be null")
         val target = target ?: throw GradleException("target cannot be null")
         val release = release ?: throw GradleException("release cannot be null")
-        val args = listOf({{quote-and-join tauri-binary-args}});
+        
+        val baseArgs = listOf({{quote-and-join tauri-binary-args}});
+        
+        // If we're using cargo as a fallback, we need to prepend "tauri" to the args
+        val args = if (executable.contains("cargo")) {
+            listOf("tauri") + baseArgs
+        } else {
+            baseArgs
+        }.toMutableList()
 
         project.exec {
             workingDir(File(project.projectDir, rootDirRel))

--- a/crates/tauri-cli/templates/mobile/android/buildSrc/src/main/kotlin/BuildTask.kt
+++ b/crates/tauri-cli/templates/mobile/android/buildSrc/src/main/kotlin/BuildTask.kt
@@ -21,12 +21,11 @@ open class BuildTask : DefaultTask() {
             runTauriCli(executable)
         } catch (e: Exception) {
             if (Os.isFamily(Os.FAMILY_WINDOWS)) {
-                // Try different Windows-specific fallbacks
+                // Try different Windows-specific extensions
                 val fallbacks = listOf(
+                    "$executable.exe",
                     "$executable.cmd",
                     "$executable.bat",
-                    "cargo.exe",
-                    "cargo"
                 )
                 
                 var lastException: Exception = e

--- a/crates/tauri-cli/templates/mobile/android/buildSrc/src/main/kotlin/BuildTask.kt
+++ b/crates/tauri-cli/templates/mobile/android/buildSrc/src/main/kotlin/BuildTask.kt
@@ -48,15 +48,7 @@ open class BuildTask : DefaultTask() {
         val rootDirRel = rootDirRel ?: throw GradleException("rootDirRel cannot be null")
         val target = target ?: throw GradleException("target cannot be null")
         val release = release ?: throw GradleException("release cannot be null")
-        
-        val baseArgs = listOf({{quote-and-join tauri-binary-args}});
-        
-        // If we're using cargo as a fallback, we need to prepend "tauri" to the args
-        val args = if (executable.contains("cargo")) {
-            listOf("tauri") + baseArgs
-        } else {
-            baseArgs
-        }.toMutableList()
+        val args = listOf({{quote-and-join tauri-binary-args}});
 
         project.exec {
             workingDir(File(project.projectDir, rootDirRel))

